File: system-test/java/src/main/java/com/optivem/eshop/dsl/driver/adapter/external/erp/ErpStubDriver.java
package com.optivem.eshop.dsl.driver.adapter.external.erp;

import com.optivem.eshop.dsl.driver.adapter.external.erp.client.ErpStubClient;
import com.optivem.eshop.dsl.driver.adapter.external.erp.client.dtos.ExtProductDetailsResponse;
import com.optivem.eshop.dsl.driver.port.external.erp.dtos.ReturnsProductRequest;
import com.optivem.eshop.dsl.driver.port.shared.dtos.ErrorResponse;
import com.optivem.eshop.dsl.common.Converter;
import com.optivem.eshop.dsl.common.Result;

/**
 * ErpStubDriver uses WireMock to stub ERP API responses.
 * This allows tests to run without a real ERP system.
 */
public class ErpStubDriver extends BaseErpDriver<ErpStubClient> {
    public ErpStubDriver(String baseUrl) {
        super(new ErpStubClient(baseUrl));
    }

    @Override
    public Result<Void, ErrorResponse> returnsProduct(ReturnsProductRequest request) {
        var extProductDetailsResponse = ExtProductDetailsResponse.builder()
                .id(request.getSku())
                .title("")
                .description("")
                .price(Converter.toBigDecimal(request.getPrice()))
                .category("")
                .brand("")
                .build();

        return client.configureGetProduct(extProductDetailsResponse)
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }
}

File: system-test/typescript/src/testkit/driver/adapter/external/erp/erp-stub-driver.ts
import type { Result } from '../../../../common/result.js';
import type { ErpErrorResponse } from '../../../port/external/erp/dtos/ErpErrorResponse.js';
import type { GetProductResponse } from '../../../port/external/erp/dtos/GetProductResponse.js';
import type { ReturnsProductRequest } from '../../../port/external/erp/dtos/ReturnsProductRequest.js';
import type { ReturnsPromotionRequest } from '../../../port/external/erp/dtos/ReturnsPromotionRequest.js';
import type { ErpDriver } from '../../../port/external/erp/erp-driver.js';
import { ErpStubClient } from './client/ErpStubClient.js';

export class ErpStubDriver implements ErpDriver {
  private readonly client: ErpStubClient;

  constructor(baseUrl: string) {
    this.client = new ErpStubClient(baseUrl);
  }

  async goToErp(): Promise<Result<void, ErpErrorResponse>> {
    return this.client.checkHealth();
  }

  async getProduct(sku: string): Promise<Result<GetProductResponse, ErpErrorResponse>> {
    return this.client.getProduct(sku);
  }

  async returnsProduct(request: ReturnsProductRequest): Promise<Result<void, ErpErrorResponse>> {
    return this.client.configureProduct(request);
  }

  async returnsPromotion(request: ReturnsPromotionRequest): Promise<Result<void, ErpErrorResponse>> {
    return this.client.configurePromotion(request);
  }

  async close(): Promise<void> {
    await this.client.close();
  }
}
